 ##  Problem Statement - 
 
We possess a dataset pertinent to a UK-based retailer during the years 2010-2011. Our objective is to conduct Market Basket Analysis on the retail data utilizing the Apriori Algorithm. Before proceeding with the analysis, preprocessing of the dataset is imperative to render it suitable for analysis.

## overview of  Dataset - 

The dataset has 8 features as given below:
InvoiceNo – Invoice number,
StockCode – Stock code of the item,
Description – Description of item,
Quantity – Quantity of item,
InvoiceDate – Data of invoice generation,
UnitPrice – Price of item per unit,
CustomerID – ID of the customer,
Country – Country of residence,

In [1]:
 ## install we will install mlxtrend Library
!pip install mlxtend



Defaulting to user installation because normal site-packages is not writeable


In [2]:
## Imoporting Libraries 

import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [3]:
## Loading the data 
## We will use the encoding as latin1 to read this file as there are some special character in this file


df = pd.read_csv("C:\\Users\\RAM\\Desktop\\Dataset\\OnlineRetail.csv",  encoding= 'unicode_escape')


In [4]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [5]:
df.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/2011 12:50       0.85     12680.0  France  
541905  12/9/2011 12:50       2.10     12680.0  France  
541906  12/9/2011 12:50       4.15     12680.0  France  
541907  12/9/2011 12:50       4.15     12680.0  France  
541908  12/9/2011 12:50       4.95     12680.0  France

In [6]:
df.shape

(541909, 8)

In [7]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [8]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [9]:
## Cleaning The data 
## we will Remove the spaces from the description column .

df['Description'] = df['Description'].str.strip()

#Drop rows without invoice number
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True)
df['InvoiceNo'] = df['InvoiceNo'].astype('str') ## convert invoice number to string

#Remove the credit transaction with invoice numbers containing 'C'
df = df[~df['InvoiceNo'].str.contains('C')]
df.head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [13]:
## after Cleanup the data 

df.shape

(532621, 8)

In [16]:
## Creating Basket 
## We are going to create a basket matrix by grouping multiple items within the same order and then unstack our DataFrame:

basket = (df[df['Country'] == "France"] ## get data for samples which have Country as France
          .groupby(['InvoiceNo', 'Description'])['Quantity'] ## group them on InvoiceNo and Description based on Quantity
          .sum() ## sum the quantity, unstack them, fill 0 to nan values.
          .unstack()
          .reset_index()
          .fillna(0)
          .set_index('InvoiceNo')) ## set the index as Invoice Number
basket.head(10)


Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                          0.0                         0.0   
536852                          0.0                         0.0   
536974                          0.0                         0.0   
537065                          0.0                         0.0   
537463                          0.0                         0.0   
537468                         24.0                         0.0   
537693                          0.0                         0.0   
537897                          0.0                         0.0   
537967                          0.0                         0.0   
538008                          0.0                         0.0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                             0.0                              0.0   
536852                             0.0                              0.0   
536974                             0.0                              0.0   
537065                             0.0                              0.0   
537463                             0.0                              0.0   
537468                             0.0                              0.0   
537693                             0.0                              0.0   
537897                             0.0                              0.0   
537967                             0.0                              0.0   
538008                             0.0                              0.0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                 0.0   
536852                                 0.0   
536974                                 0.0   
537065                                 0.0   
537463                                 0.0   
537468                                 0.0   
537693                                 0.0   
537897                                 0.0   
537967                                 0.0   
538008                                 0.0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                       0.0                          0.0   
536852                                       0.0                          0.0   
536974                                       0.0                          0.0   
537065                                       0.0                          0.0   
537463                                       0.0                          0.0   
537468                                       0.0                          0.0   
537693                                       0.0                          0.0   
537897                                       0.0                          0.0   
537967                                       0.0                          0.0   
538008                                       0.0                          0.0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                             0.0                                 0.0   
536852                             0.0                                 0.0   
536974                             0.0                                 0.0   
537065                             0.0                                 0.0   
537463                             0.0                                 0.0   
537468                             0.0                                 0.0   
537693                             0.0                                 0.0   
537897                             0.0                            

In [17]:
basket.shape

(392, 1563)

## There are a lot of zeros in the data but we also need to make sure any positive values are converted to a 1 and anything less the 0 is set to 0.

In [18]:
#below function converts a values < 0 to 0 and value greater than equal 1 to 1.  
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
#apply function to data using applymap.
basket_sets = basket.applymap(encode_units)

In [19]:
basket_sets.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                            0                           0   
536852                            0                           0   
536974                            0                           0   
537065                            0                           0   
537463                            0                           0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                               0                                0   
536852                               0                                0   
536974                               0                                0   
537065                               0                                0   
537463                               0                                0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                         0                            0   
536852                                         0                            0   
536974                                         0                            0   
537065                                         0                            0   
537463                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                               0                                   0   
536852                               0                                   0   
536974                               0                                   0   
537065                               0                                   0   
537463                               0                                   0   

Description  12 PENCILS TALL TUBE WOODLAND  ...  WRAP VINTAGE PETALS  DESIGN  \
InvoiceNo                                   ...                                
536370                                   0  ...                            0   
536852                                   0  ...                            0   
536974                                   0  ...                            0   
537065                                   0  ...                            0   
537463                                   0  ...                            0   

Description  YELLOW COAT RACK PARIS FASHION  YELLOW GIANT GARDEN THERMOMETER  \
InvoiceNo                                                                      
536370                                    0                                0   
536852                                    0                                0   
536974                                    0                                0   
537065                                    0                                0   
537463                                    0                                0   

Description  YELLOW SHARK HELICOPTER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                         
536370                             0                          0   
536852                             0                          0   
536974                             0                          0   
537065                             0                          0   
537463                             0      

## Generating Frequent items sets 
that the data is structured properly, we can generate frequent item sets that have a support of at least 7% (this number was chosen so that I could get enough useful examples)

In [20]:
frequent_itemsets = apriori(basket_sets, min_support=0.07, use_colnames=True)

C:\Users\RAM\AppData\Roaming\Python\Python310\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


## Generating Association rules
We will generate association rules with their corresponding support, confidence, and lift

In [21]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
 
rules.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2    (ALARM CLOCK BAKELIKE RED)  (ALARM CLOCK BAKELIKE GREEN)   
3  (ALARM CLOCK BAKELIKE GREEN)    (ALARM CLOCK BAKELIKE RED)   
4                     (POSTAGE)  (ALARM CLOCK BAKELIKE GREEN)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.096939            0.102041  0.073980    0.763158  7.478947   
1            0.102041            0.096939  0.073980    0.725000  7.478947   
2            0.094388            0.096939  0.079082    0.837838  8.642959   
3            0.096939            0.094388  0.079082    0.815789  8.642959   
4            0.765306            0.096939  0.084184    0.110000  1.134737   

   leverage  conviction  zhangs_metric  
0  0.064088    3.791383       0.959283  
1  0.064088    3.283859       0.964734  
2  0.069932    5.568878       0.976465  
3  0.069932    4.916181       0.979224  
4  0.009996    1.014676       0.505929

## Filtering Rules with high Confidence and Lift

As we already know, the greater the lift ratio, the more significant the association between the items. Also, the higher the confidence, the more reliable are the rules.

So, we are looking for rules with high confidence (>=0.8) and high lift (>=6). For this we filter the records as shown

In [23]:
## Filtering out the values 
rules[(rules['lift'] >= 6) & (rules['confidence'] >= 0.8)]


antecedents  \
2                           (ALARM CLOCK BAKELIKE RED)   
3                         (ALARM CLOCK BAKELIKE GREEN)   
74                     (SET/6 RED SPOTTY PAPER PLATES)   
76                     (SET/6 RED SPOTTY PAPER PLATES)   
77                       (SET/6 RED SPOTTY PAPER CUPS)   
78                 (ALARM CLOCK BAKELIKE RED, POSTAGE)   
80               (POSTAGE, ALARM CLOCK BAKELIKE GREEN)   
115           (SET/6 RED SPOTTY PAPER PLATES, POSTAGE)   
116             (SET/6 RED SPOTTY PAPER CUPS, POSTAGE)   
117                    (SET/6 RED SPOTTY PAPER PLATES)   
120  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   
121  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   
122  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
128  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   
129  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   

                                consequents  antecedent support  \
2              (ALARM CLOCK BAKELIKE GREEN)            0.094388   
3                (ALARM CLOCK BAKELIKE RED)            0.096939   
74     (SET/20 RED RETROSPOT PAPER NAPKINS)            0.127551   
76            (SET/6 RED SPOTTY PAPER CUPS)            0.127551   
77          (SET/6 RED SPOTTY PAPER PLATES)            0.137755   
78             (ALARM CLOCK BAKELIKE GREEN)            0.086735   
80               (ALARM CLOCK BAKELIKE RED)            0.084184   
115           (SET/6 RED SPOTTY PAPER CUPS)            0.107143   
116         (SET/6 RED SPOTTY PAPER PLATES)            0.117347   
117  (SET/6 RED SPOTTY PAPER CUPS, POSTAGE)            0.127551   
120    (SET/20 RED RETROSPOT PAPER NAPKINS)            0.122449   
121           (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
122         (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
128           (SET/6 RED SPOTTY PAPER CUPS)            0.084184   
129         (SET/6 RED SPOTTY PAPER PLATES)            0.084184   

     consequent support   support  confidence      lift  leverage  conviction  \
2              0.096939  0.079082    0.837838  8.642959  0.069932    5.568878   
3              0.094388  0.079082    0.815789  8.642959  0.069932    4.916181   
74             0.132653  0.102041    0.800000  6.030769  0.085121    4.336735   
76             0.137755  0.122449    0.960000  6.968889  0.104878   21.556122   
77             0.127551  0.122449    0.888889  6.968889  0.104878    7.852041   
78             0.096939  0.071429    0.823529  8.495356  0.063021    5.117347   
80             0.094388  0.071429    0.848485  8.989353  0.063483    5.977041   
115            0.137755  0.102041    0.952381  6.913580  0.087281   18.107143   
116            0.127551  0.102041    0.869565  6.817391  0.087073    6.688776   
117            0.117347  0.102041    0.800000  6.817391  0.087073    4.413265   
120            0.132653  0.099490    0.812500  6.125000  0.083247    4.625850   
121            0.137755  0.099490    0.975000  7.077778  0.085433   34.489796   
122            0.127551  0.099490    0.975000  7.644000  0.086474   34.897959   
128            0.137755  0.081633    0.969697  7.039282  0.070036   28.454082   
129            0.127551  0.081633    0.969697  7.602424  0.070895   28.790816   

     zhangs_metric  
2         0.976465  
3         0.979224  
74        0.956140  
76        0.981725  
77        0.993343  
78        0.966081  
80        0.970454  
115       0.958000  
116       0.966763  
117       0.978070  
120       0.953488  
121       0.956294  
122       0.967949  
128       0.936804  
129       0.948294

## output 

So, we have our antecedents and consequent items along with their parameter values.

The result is giving us a lot of information about item grouping such as customers are 84% likely to buy the green alarm clock if they have already bought the red alarm clock!

## Conclusion 

Association Rule Mining through the Apriori algorithm is a widely known technique in machine learning to perform Market Basket Analysis. This analyzing process is used to discover the association between products in a manner that will be helpful for retailers or marketers in developing useful marketing strategies and business plans